upb/mini_table.c

```diff
@@ -0,0 +1,937 @@
+  }
+
+  upb_MtDecoder_CheckOutOfMemory(&decoder, decoder.table);
+
+  decoder.table->size = 0;
+  decoder.table->field_count = 0;
+  decoder.table->ext = kUpb_ExtMode_NonExtendable;
+  decoder.table->dense_below = 0;
+  decoder.table->table_mask = 0;
+  decoder.table->required_count = 0;
+
+  upb_MtDecoder_ParseMessage(&decoder, data, len);
+  upb_MtDecoder_AssignHasbits(decoder.table);
+  upb_MtDecoder_SortLayoutItems(&decoder);
+  upb_MtDecoder_AssignOffsets(&decoder);
+
+done:
+  *buf = decoder.vec.data;
+  *buf_size = decoder.vec.capacity / sizeof(*decoder.vec.data);
+  return decoder.table;
+}
+
+upb_MiniTable* upb_MiniTable_BuildMessageSet(upb_MiniTablePlatform platform,
+                                             upb_Arena* arena) {
+  upb_MiniTable* ret = upb_Arena_Malloc(arena, sizeof(*ret));
+  if (!ret) return NULL;
+
+  ret->size = 0;
+  ret->field_count = 0;
+  ret->ext = kUpb_ExtMode_IsMessageSet;
+  ret->dense_below = 0;
+  ret->table_mask = 0;
+  ret->required_count = 0;
+  return ret;
+}
+
+upb_MiniTable* upb_MiniTable_BuildMapEntry(upb_FieldType key_type,
+                                           upb_FieldType value_type,
+                                           upb_MiniTablePlatform platform,
+                                           upb_Arena* arena) {
+  upb_MiniTable* ret = upb_Arena_Malloc(arena, sizeof(*ret));
+  upb_MiniTable_Field* fields = upb_Arena_Malloc(arena, sizeof(*fields) * 2);
+  if (!ret || !fields) return NULL;
+
+  upb_MiniTable_Sub* subs = NULL;
+  if (value_type == kUpb_FieldType_Message || value_type == kUpb_FieldType_Group) {
+    subs = upb_Arena_Malloc(arena, sizeof(*subs));
+    if (!subs) return NULL;
+  }
+
+  size_t field_size =
+      upb_MtDecoder_SizeOfRep(kUpb_FieldRep_StringView, platform);
+
+  fields[0].number = 1;
+  fields[1].number = 2;
+  fields[0].mode = kUpb_FieldMode_Scalar;
+  fields[1].mode = kUpb_FieldMode_Scalar;
+  fields[0].presence = 0;
+  fields[1].presence = 0;
+  fields[0].descriptortype = key_type;
+  fields[1].descriptortype = value_type;
+  fields[0].offset = 0;
+  fields[1].offset = field_size;
+  fields[1].submsg_index = 0;
+
+  ret->size = UPB_ALIGN_UP(2 * field_size, 8);
+  ret->field_count = 2;
+  ret->ext = kUpb_ExtMode_NonExtendable | kUpb_ExtMode_IsMapEntry;
+  ret->dense_below = 2;
+  ret->table_mask = 0;
+  ret->required_count = 0;
+  ret->subs = subs;
+  ret->fields = fields;
+  return ret;
+}
+
+upb_MiniTable_Extension* upb_MiniTable_BuildExtensions(const char* data,
+                                                       size_t len,
+                                                       size_t* ext_count,
+                                                       upb_Arena* arena,
+                                                       upb_Status* status) {
+  upb_MtDecoder decoder = {
+      .arena = arena,
+      .status = status,
+      .table = NULL,
+  };
+
+  upb_MiniTable_Extension* exts;
+
+  if (UPB_SETJMP(decoder.err)) {
+    exts = NULL;
```

Comment:
can you move this to the variable declaration?

Reply:
It would have to be marked `volatile` to have a guarantee of being preserved by longjmp(). I'd prefer not to.